File: core/src/nl/t64/game/rpg/components/party/StatContainer.java
package nl.t64.game.rpg.components.party;

import com.fasterxml.jackson.annotation.JsonCreator;

import java.beans.ConstructorProperties;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Collectors;


class StatContainer {

    private static final int NUMBER_OF_STAT_SLOTS = 7;

    private Level level;
    private final Map<String, StatItem> stats;

    private StatContainer() {
        this.stats = new HashMap<>(NUMBER_OF_STAT_SLOTS);
    }

    @JsonCreator
    @ConstructorProperties({
            "level", "intelligence", "willpower", "dexterity",
            "agility", "endurance", "strength", "stamina"})
    private StatContainer(int lvl, int inl, int wil, int dex, int agi, int edu, int str, int sta) {
        this();
        this.level = new Level(lvl);
        this.stats.put(StatItemId.INTELLIGENCE.name(), new Intelligence(inl));
        this.stats.put(StatItemId.WILLPOWER.name(), new Willpower(wil));
        this.stats.put(StatItemId.DEXTERITY.name(), new Dexterity(dex));
        this.stats.put(StatItemId.AGILITY.name(), new Agility(agi));
        this.stats.put(StatItemId.ENDURANCE.name(), new Endurance(edu));
        this.stats.put(StatItemId.STRENGTH.name(), new Strength(str));
        this.stats.put(StatItemId.STAMINA.name(), new Stamina(sta));
    }

    int getXpNeededForNextLevel() {
        return level.getXpNeededForNextLevel();
    }

    int getXpDeltaBetweenLevels() {
        return level.getXpDeltaBetweenLevels();
    }

    void gainXp(int amount, Consumer<Boolean> hasGainedLevel) {
        level.gainXp(amount, hasGainedLevel);
    }

    int getTotalXp() {
        return level.totalXp;
    }

    int getXpToInvest() {
        return level.xpToInvest;
    }

    int getLevel() {
        return level.rank;
    }

    Map<String, Integer> getAllHpStats() {
        return Map.of("lvlRank", level.rank,
                      "lvlVari", level.variable,
                      "staRank", stats.get(StatItemId.STAMINA.name()).rank,
                      "staVari", stats.get(StatItemId.STAMINA.name()).variable,
                      "eduRank", stats.get(StatItemId.ENDURANCE.name()).rank,
                      "eduVari", stats.get(StatItemId.ENDURANCE.name()).variable,
                      "eduBon", stats.get(StatItemId.ENDURANCE.name()).bonus);
    }

    int getMaximumHp() {
        return level.rank
               + stats.get(StatItemId.STAMINA.name()).rank
               + stats.get(StatItemId.ENDURANCE.name()).rank
               + stats.get(StatItemId.ENDURANCE.name()).bonus;
    }

    int getCurrentHp() {
        return level.variable
               + stats.get(StatItemId.STAMINA.name()).variable
               + stats.get(StatItemId.ENDURANCE.name()).variable
               + stats.get(StatItemId.ENDURANCE.name()).bonus;
    }

    StatItem getById(StatItemId statItemId) {
        return stats.get(statItemId.name());
    }

    List<StatItem> getAll() {
        return Arrays.stream(StatItemId.values())
                     .map(statItemId -> stats.get(statItemId.name()))
                     .collect(Collectors.toUnmodifiableList());
    }

}

File: core/src/nl/t64/game/rpg/components/party/stats/StatContainer.kt
package nl.t64.game.rpg.components.party.stats

import java.beans.ConstructorProperties


private const val NUMBER_OF_STAT_SLOTS = 7

class StatContainer() {

    private lateinit var level: Level
    private val stats: MutableMap<String, StatItem> = HashMap(NUMBER_OF_STAT_SLOTS)

    @ConstructorProperties(
        "level", "intelligence", "willpower", "dexterity", "agility", "endurance", "strength", "stamina")
    constructor(lvl: Int, inl: Int, wil: Int, dex: Int, agi: Int, edu: Int, str: Int, sta: Int) : this() {
        this.level = Level(lvl)
        this.stats[StatItemId.INTELLIGENCE.name] = Intelligence(inl)
        this.stats[StatItemId.WILLPOWER.name] = Willpower(wil)
        this.stats[StatItemId.DEXTERITY.name] = Dexterity(dex)
        this.stats[StatItemId.AGILITY.name] = Agility(agi)
        this.stats[StatItemId.ENDURANCE.name] = Endurance(edu)
        this.stats[StatItemId.STRENGTH.name] = Strength(str)
        this.stats[StatItemId.STAMINA.name] = Stamina(sta)
    }

    fun getXpNeededForNextLevel(): Int = level.getXpNeededForNextLevel()
    fun getXpDeltaBetweenLevels(): Int = level.getXpDeltaBetweenLevels()
    fun getTotalXp(): Int = level.totalXp
    fun getXpToInvest(): Int = level.xpToInvest
    fun getLevel(): Int = level.rank

    fun gainXp(amount: Int, hasGainedLevel: (Boolean) -> Unit) {
        level.gainXp(amount, hasGainedLevel)
    }

    fun getAllHpStats(): Map<String, Int> {
        return mapOf(Pair("lvlRank", level.rank),
                     Pair("lvlVari", level.variable),
                     Pair("staRank", stats[StatItemId.STAMINA.name]!!.rank),
                     Pair("staVari", stats[StatItemId.STAMINA.name]!!.variable),
                     Pair("eduRank", stats[StatItemId.ENDURANCE.name]!!.rank),
                     Pair("eduVari", stats[StatItemId.ENDURANCE.name]!!.variable),
                     Pair("eduBon", stats[StatItemId.ENDURANCE.name]!!.bonus))
    }

    fun getMaximumHp(): Int {
        return (level.rank
                + stats[StatItemId.STAMINA.name]!!.rank
                + stats[StatItemId.ENDURANCE.name]!!.rank
                + stats[StatItemId.ENDURANCE.name]!!.bonus)
    }

    fun getCurrentHp(): Int {
        return (level.variable
                + stats[StatItemId.STAMINA.name]!!.variable
                + stats[StatItemId.ENDURANCE.name]!!.variable
                + stats[StatItemId.ENDURANCE.name]!!.bonus)
    }

    fun getById(statItemId: StatItemId): StatItem {
        return stats[statItemId.name]!!
    }

    fun getAll(): List<StatItem> {
        return StatItemId.values().map { stats[it.name]!! }
    }

    fun takeDamage(damage: Int) {
        level.takeDamage(damage)?.let { it1 ->
            (stats[StatItemId.STAMINA.name] as Stamina).takeDamage(it1)?.let { it2 ->
                (stats[StatItemId.ENDURANCE.name] as Endurance).takeDamage(it2)
            }
        }
    }

    fun getInflictDamageStaminaPenalty(): Int {
        return (stats[StatItemId.STAMINA.name] as Stamina).getInflictDamagePenalty()
    }

}
